AWS cloud provider

abutcher opened the pull request:
Adds AWS cloud provider support to openshift-ansible and modifies inventory variables used for the openstack cloud provider.

* AWS access and secret keys may be omitted if IAM profiles are in use.
* AWS region used in cloud provider config is pulled from instance metadata.

```
# Cloud Provider Configuration
#
# Note: You may make use of environment variables rather than store
# sensitive configuration within the ansible inventory.
# For example:
#openshift_cloudprovider_aws_access_key="{{ lookup('env','AWS_ACCESS_KEY_ID') }}"
#openshift_cloudprovider_aws_secret_key="{{ lookup('env','AWS_SECRET_ACCESS_KEY') }}"
#
# AWS
#openshift_cloudprovider_kind=aws
# Note: IAM profiles may be used instead of storing API credentials on disk.
#openshift_cloudprovider_aws_access_key=aws_access_key_id
#openshift_cloudprovider_aws_secret_key=aws_secret_access_key
#
# Openstack
#openshift_cloudprovider_kind=openstack
#openshift_cloudprovider_openstack_auth_url=http://openstack.example.com:35357/v2.0/
#openshift_cloudprovider_openstack_username=username
#openshift_cloudprovider_openstack_password=password
#openshift_cloudprovider_openstack_tenand_id=tenant_id
#openshift_cloudprovider_openstack_tenant_name=tenant_name
#openshift_cloudprovider_openstack_region=region
#openshift_cloudprovider_openstack_lb_subnet_id=subnet_id
```

Resolves #1522

detiber commented on inventory/byo/hosts.origin.example:
```diff
@@ -95,6 +95,22 @@ openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true',
+# AWS
+#openshift_cloud_provider=aws
+#openshift_cloud_provider_aws_access_key=AWS_ACCESS_KEY_ID
+#openshift_cloud_provider_aws_secret_key=AWS_SECRET_ACCESS_KEY
```
We'll probably want to document how to set these values using the lookup('env') so that users *can* use their environment variables if they wish.

abutcher replied:
Agreed. This also needs a comment regarding IAM profiles.

detiber commented on roles/openshift_cloud_provider/tasks/aws.yml:
```diff
@@ -0,0 +1,18 @@
+    role: aws
+    local_facts:
+      access_key: "{{ openshift_cloud_provider_aws_access_key | default(None) }}"
+      secret_key: "{{ openshift_cloud_provider_aws_secret_key | default(None) }}"
```
Instead of using local facts here, should we instead use openshift_env, and store these under a 'cloud_provider' role (`openshift.cloud_provider.aws.access_key`)?

abutcher replied:
I considered this but that would end up being `openshift.cloud.provider.aws.access.key` which is fine by me, nobody is going to see it...

detiber replied:
I just hate to introduce a 'role' for each cloud provider.

abutcher replied:
What about a `cloudprovider` one word role?

abutcher replied:
`cloud` seems too generic to me

detiber replied:
:+1: to cloudprovider

detiber commented on roles/openshift_master/templates/native-cluster/atomic-openshift-master-controllers.j2:
```diff
@@ -4,6 +4,11 @@ CONFIG_FILE={{ openshift_master_config_file }}
+AWS_ACCESS_KEY_ID={{ openshift.aws.access_key }}
+AWS_SECRET_ACCESS_KEY={{ openshift.aws.secret_key }}
+{% endif %}
+
```
Man, we really need to consolidate this template...

detiber commented on roles/openshift_common/meta/main.yml:
```diff
@@ -12,7 +12,7 @@ galaxy_info:
+- role: os_firewall
+- role: openshift_facts
+- role: openshift_repos
+- role: openshift_cloud_provider
```
Should openshift_cloud_provider be a dependency here, or on the master/node roles instead? The opeshift_cli role for example wouldn't need cloud_provider config...

abutcher replied:
Depending on node/master sounds way better to me.

dak1n1 commented on inventory/byo/hosts.ose.example:
```diff
@@ -91,6 +91,22 @@ openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true',
 #openshift_master_identity_providers=[{'name': 'my_ldap_provider', 'challenge': 'true', 'login': 'true', 'kind': 'LDAPPasswordIdentityProvider', 'attributes': {'id': ['dn'], 'email': ['mail'], 'name': ['cn'], 'preferredUsername': ['uid']}, 'bindDN': '', 'bindPassword': '', 'ca': '', 'insecure': 'false', 'url': 'ldap://ldap.example.com:389/ou=users,dc=example,dc=com?uid'}]
 
+# Cloud provider facts
+#
```
This looks great! Would I also put my zone in here? Just trying to get a head start on the next build... :)

abutcher replied:
We're already pulling the zone info out of metadata so I'm just reusing
here.

On Monday, March 21, 2016, Stefanie Forrester <notifications@github.com>
wrote:

> In inventory/byo/hosts.ose.example
> <https://github.com/openshift/openshift-ansible/pull/1634#discussion_r56908288>
> :
>
> > @@ -91,6 +91,22 @@ openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true',
> >  # LDAP auth
> >  #openshift_master_identity_providers=[{'name': 'my_ldap_provider', 'challenge': 'true', 'login': 'true', 'kind': 'LDAPPasswordIdentityProvider', 'attributes': {'id': ['dn'], 'email': ['mail'], 'name': ['cn'], 'preferredUsername': ['uid']}, 'bindDN': '', 'bindPassword': '', 'ca': '', 'insecure': 'false', 'url': 'ldap://ldap.example.com:389/ou=users,dc=example,dc=com?uid'}]
> >
> > +# Cloud provider facts
> > +#
>
> This looks great! Would I also put my zone in here? Just trying to get a
> head start on the next build... :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/openshift-ansible/pull/1634/files/82c41633c931e6a3821867b5c5adb385b260c892#r56908288>
>

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

detiber commented on roles/openshift_cloud_provider/tasks/main.yml:
```diff
@@ -1,3 +1,14 @@
+  when: "'cloudprovider' in openshift and 'kind' in openshift.cloudprovider and openshift.cloudprovider.kind == 'openstack' and 'openstack' in openshift.cloudprovider"
+
+- include: aws.yml
+  when: "'cloudprovider' in openshift and 'kind' in openshift.cloudprovider and openshift.cloudprovider.kind == 'aws'"
```
Does this not work?
```
when: openshift.cloudprovider.kind | default('') == 'aws'
```

abutcher replied:
If no `openshift_cloudprovider_` vars are supplied then we have to perform all these checks :(

abutcher replied:
Although, my defaults for cloudprovider should leave us with at least a kind of None.

detiber commented on roles/openshift_facts/library/openshift_facts.py:
```diff
@@ -337,9 +337,14 @@ def normalize_provider_facts(provider, metadata):
             facts[arg] = {'cloud-provider': [provider],
                           'cloud-config': ['/etc/cloud.conf']}
 
+    if os.path.exists('/etc/aws/aws.conf'):
```
These should probably be dependent on the existence of openshift.cloudprovider.kind instead of testing for the file on disk...

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
@detiber PTAL I added some functions to openshift_facts that we can build on for the current args vars

@lebauce Could you take a look also? I'm doing quite a bit of refactoring around the variables for each provider which would mean that environment vars for the openstack provider should be set in the inventory. I added an example the inventories for setting these.

detiber commented on inventory/byo/hosts.aep.example:
```diff
@@ -90,6 +90,30 @@ openshift_master_identity_providers=[{'name': 'htpasswd_auth', 'login': 'true',
+#openshift_cloudprovider_aws_secret_key="{{ lookup('env','AWS_SECRET_ACCESS_KEY') }}"
+#
+# AWS
+# Note: IAM profiles may be used instead of storing API credentials on disk.
```
Should this note be moved down below openshift_cloud_provider_kind, or otherwise suggest that openshift_cloudprovider_kind still needs to be specified?

abutcher replied:
Sure thing.

detiber commented on roles/openshift_cloud_provider/meta/main.yml:
```diff
@@ -0,0 +1,15 @@
+---
+galaxy_info:
+  author: Sylvain Baubeau
```
Not a huge deal currently, but we might need to find a way to attribute authorship to multiple people, and specify who is the maintainer somehow.

detiber commented on roles/openshift_cloud_provider/tasks/main.yml:
```diff
@@ -1,3 +1,17 @@
+
 - include: openstack.yml
-  when: "openshift_cloud_provider is defined and openshift_cloud_provider == 'openstack' and 'provider' in openshift and openshift.provider.name == 'openstack'"
+  when: "'cloudprovider' in openshift and 'kind' in openshift.cloudprovier and openshift.cloudprovider.kind == 'openstack' and 'openstack' in openshift.cloudprovider"
```
too bad we can't use the 'fileglob' lookup and loop here...  maybe in ansible 2.x with a block.

detiber replied:
I think you told me there was a reason that we couldn't just use openshift.cloudprovider.kind | default('') == 'openstack'

As an aside, I see `openshift.cloudprovier` in the above line.

abutcher replied:
That default will still try to look in the dictionary so we have to test the keys

detiber replied:
ah, carry on then.

detiber commented on roles/openshift_facts/library/openshift_facts.py:
```diff
@@ -918,6 +914,51 @@ def get_current_config(facts):
+        if 'cloudprovider' in facts:
+            if facts['cloudprovider']['kind'] == 'aws':
+                kubelet_args['cloud-provider'] = 'aws'
+                kubelet_args['cloud-config'] = ['/etc/aws/aws.conf']
```
Should we actually put this in /etc/aws or should we put it in /etc/origin/cloudprovider/aws.conf ?

Same with the openstack file below...  /etc/origin/cloudprovider/openstack.conf?

abutcher replied:
Our docs use the locations mentioned here but both could be updated easily enough

detiber commented:
Some minor nits picked.

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
@detiber PTAL - I moved the config file locations and I also added `no_log` to the cloud provider facts to suppress outputting hostvars/hostvars[hostname] for each host.

detiber commented on roles/openshift_cloud_provider/tasks/main.yml:
```diff
@@ -6,12 +6,19 @@
+  file:
+    path: "{{ openshift.common.config_base }}/cloudprovider"
+    state: directory
+  when: "'cloudprovider' in openshift and 'kind' in openshift.cloudprovider and openshift.cloudprovider.kind != None"
```
I wonder if it makes sense to do something like this in vars/main.yml:
```
has_cloudprovider: "{{ 'cloudprovider' in openshift and 'kind' in openshift.cloudprovider and openshift.cloudprovider.kind is not None }}"
cloudprovider_is_aws: "{{ has_cloudprovider | bool and openshift.cloudprovider.kind == 'aws' }}"
cloudprovider_is_openstack: "{{ has_cloudprovider | bool and openshift.cloudprovider.kind == 'openstack' }}"
```

abutcher replied:
Makes sense to me.

abutcher commented:
aos-ci-test

abutcher commented:
aos-ci-test

abutcher commented:
openstack testing looks good on my end

```
Mar 31 11:02:35 abutcher-test.localdomain atomic-openshift-master[13067]: I0331 11:02:35.993149   13067 master_config.go:129] Successfully initialized cloud provider: "openstack" from the config file: "/etc/origin/cloudprovider/openstack.conf"
```

```
{"cloudprovider": {"openstack": {"username": "redacted", "tenant_name": "redacted", "region": "redacted", "password": "redacted", "auth_url": "http://redacted:5000/v2.0"}, "kind": "openstack"}}
```

```
...
kubernetesMasterConfig:
  apiServerArguments:
    cloud-config:
    - /etc/origin/cloudprovider/openstack.conf
    cloud-provider:
    - openstack
  controllerArguments:
    cloud-config:
    - /etc/origin/cloudprovider/openstack.conf
    cloud-provider:
    - openstack
...
```

detiber commented:
@abutcher lgtm we'll need to rope in @adellape to start on the docs work for this feature.

@brenton

brenton commented:
@abutcher, @detiber, amazing work on this!
